File: Opaline.Core/Playback/Sabr/UmpReader.cs
namespace Opaline.Core.Playback.Sabr;

public readonly record struct UmpPart(int Type, byte[] Payload);

public enum UmpPartType
{
    MediaHeader = 20,
    Media = 21,
    MediaEnd = 22,
    NextRequestPolicy = 35,
    FormatInitializationMetadata = 42,
    SabrRedirect = 43,
    SabrError = 44,
    SabrSeek = 45,
    ReloadPlayerResponse = 46,
    EndOfTrack = 62,
}

/// <summary>Incremental UMP (type, size, payload) reader — iOS UMPReader.</summary>
public sealed class UmpReader
{
    private readonly List<byte> _buffer = new();

    public void Append(ReadOnlySpan<byte> chunk)
    {
        foreach (var b in chunk) _buffer.Add(b);
    }

    public void Append(byte[] chunk) => Append(chunk.AsSpan());

    public List<UmpPart> ReadParts()
    {
        var parts = new List<UmpPart>();
        var data = _buffer.ToArray();
        var offset = 0;
        while (true)
        {
            if (!SabrProtobuf.TryReadUmpVarint(data, offset, out var type, out var afterType))
                break;
            if (!SabrProtobuf.TryReadUmpVarint(data, afterType, out var size, out var afterSize))
                break;
            if (size < 0 || afterSize + size > data.Length)
                break;
            var payload = new byte[size];
            Buffer.BlockCopy(data, afterSize, payload, 0, size);
            parts.Add(new UmpPart(type, payload));
            offset = afterSize + size;
        }
        if (offset > 0)
            _buffer.RemoveRange(0, offset);
        return parts;
    }
}

File: ios/Opaline/Core/Playback/SABR/UMPReader.swift
import Foundation

/// One UMP part: a type tag and its payload.
struct UMPPart {
    let type: Int
    let payload: Data
}

/// Part types we act on. The server sends plenty more; everything unlisted is
/// skipped, which is the documented way to stay forward-compatible with UMP.
/// Numbers are YouTube's own, cross-checked against two independent
/// implementations of this protocol (SmartTube's `UMPPartId`, LuanRT's
/// `googlevideo`) — an earlier guess had 43-45 and 57 shifted by one, which
/// made a server seek look like an expired player response.
enum UMPPartType: Int {
    case mediaHeader = 20
    case media = 21
    case mediaEnd = 22
    /// Readahead targets, backoff and the playback cookie — the server pacing
    /// the client. See `NextRequestPolicy` in googlevideo's protos.
    case nextRequestPolicy = 35
    /// Where a format ends: last segment number and end time.
    case formatInitializationMetadata = 42
    case sabrRedirect = 43
    case sabrError = 44
    case sabrSeek = 45
    case reloadPlayerResponse = 46
    /// Bookkeeping the server attaches to ordinary responses. Listed only so
    /// logs name them instead of printing bare numbers — a response carrying
    /// nothing but these was once mistaken for the end of the stream.
    case playbackStartPolicy = 47
    case requestIdentifier = 52
    case requestCancellationPolicy = 53
    case sabrContextUpdate = 57
    case streamProtectionStatus = 58
    case endOfTrack = 62
}

/// Incremental reader for UMP's length-prefixed framing.
///
/// A response is a flat run of `(type, size, payload)` triples, both numbers in
/// UMP's own varint — unrelated to protobuf's: the byte count is encoded in the
/// leading bits of the first byte, and the remaining bytes are little-endian.
final class UMPReader {
    private var buffer = Data()

    /// UMP varint: byte count comes from the high bits of the first byte, the
    /// low bits carry the value's head, trailing bytes are little-endian.
    static func readVarint(_ data: Data, _ offset: Int) -> (Int, Int)? {
        guard offset >= 0, offset < data.count else {
            return nil
        }
        let first = Int(data[data.startIndex + offset])
        let length: Int
        switch first {
        case ..<128:
            length = 1
        case ..<192:
            length = 2
        case ..<224:
            length = 3
        case ..<240:
            length = 4
        default:
            length = 5
        }
        guard offset + length <= data.count else {
            return nil
        }
        // The 5-byte form drops the first byte entirely and is a plain uint32.
        guard length < 5 else {
            return (Int(readLittleEndian(data, offset + 1, count: 4)), offset + 5)
        }
        let head = first & (0xFF >> length)
        let tail = readLittleEndian(data, offset + 1, count: length - 1)
        return (head + Int(tail) << (8 - length), offset + length)
    }

    private static func readLittleEndian(
        _ data: Data,
        _ offset: Int,
        count: Int
    ) -> UInt64 {
        var value: UInt64 = 0
        for index in 0..<count {
            value |= UInt64(data[data.startIndex + offset + index]) << (8 * index)
        }
        return value
    }

    func append(_ chunk: Data) {
        buffer.append(chunk)
    }

    /// Returns every part fully present in the buffer and consumes it. A part
    /// split across chunk boundaries stays buffered until the rest arrives.
    func readParts() -> [UMPPart] {
        var parts: [UMPPart] = []
        var offset = 0
        while true {
            guard let (type, afterType) = Self.readVarint(buffer, offset),
                  let (size, afterSize) = Self.readVarint(buffer, afterType),
                  size >= 0, afterSize + size <= buffer.count
            else {
                break
            }
            // Offsets here are relative to the buffer's contents, and Data is
            // not always indexed from zero once bytes have been dropped off
            // the front — so every subrange is built off startIndex.
            let base = buffer.startIndex + afterSize
            parts.append(UMPPart(type: type, payload: buffer.subdata(in: base..<(base + size))))
            offset = afterSize + size
        }
        if offset > 0 {
            buffer.removeSubrange(buffer.startIndex..<(buffer.startIndex + offset))
        }
        return parts
    }
}
